File: src/data/pulse/metro.js
const metro = {
  id: 'metro',
  name: '地铁线路流量',
  description: '广州地铁各线路客流强度与主要站点日均客流量',
  mapData: {
    type: 'heatmap',
    points: [
      // 1号线
      { lat: 23.1280, lng: 113.2480, value: 85000, name: '西塱', district: '荔湾', line: '1号线' },
      { lat: 23.1200, lng: 113.2400, value: 92000, name: '陈家祠', district: '荔湾', line: '1号线' },
      { lat: 23.1280, lng: 113.2580, value: 128000, name: '公园前', district: '越秀', line: '1号线/2号线' },
      { lat: 23.1320, lng: 113.2750, value: 115000, name: '东山口', district: '越秀', line: '1号线/6号线' },
      { lat: 23.1350, lng: 113.3200, value: 155000, name: '体育西路', district: '天河', line: '1号线/3号线' },
      { lat: 23.1255, lng: 113.3320, value: 118000, name: '珠江新城', district: '天河', line: '3号线/5号线' },
      // 2号线
      { lat: 23.1600, lng: 113.2700, value: 72000, name: '白云公园', district: '白云', line: '2号线' },
      { lat: 23.1300, lng: 113.2650, value: 95000, name: '纪念堂', district: '越秀', line: '2号线' },
      { lat: 23.1190, lng: 113.2750, value: 105000, name: '海珠广场', district: '越秀', line: '2号线/6号线' },
      { lat: 23.0900, lng: 113.2900, value: 88000, name: '江南西', district: '海珠', line: '2号线' },
      { lat: 22.9500, lng: 113.3200, value: 65000, name: '市桥', district: '番禺', line: '3号线' },
      { lat: 23.1700, lng: 113.2500, value: 78000, name: '三元里', district: '白云', line: '2号线' },
      // 3号线（最繁忙线路）
      { lat: 23.1350, lng: 113.3450, value: 142000, name: '体育中心', district: '天河', line: '3号线' },
      { lat: 23.1200, lng: 113.3550, value: 96000, name: '车陂南', district: '天河', line: '3号线/4号线' },
      { lat: 22.9400, lng: 113.3200, value: 72000, name: '番禺广场', district: '番禺', line: '3号线' },
      { lat: 23.1300, lng: 113.3200, value: 138000, name: '林和西', district: '天河', line: '3号线' },
      { lat: 23.1250, lng: 113.3150, value: 125000, name: '广州东站', district: '天河', line: '1号线/3号线' },
      { lat: 23.1050, lng: 113.3400, value: 108000, name: '客村', district: '海珠', line: '3号线/8号线' },
      // 5号线
      { lat: 23.1250, lng: 113.3380, value: 112000, name: '猎德', district: '天河', line: '5号线' },
      { lat: 23.1180, lng: 113.3380, value: 98000, name: '员村', district: '天河', line: '5号线/21号线' },
      { lat: 23.1000, lng: 113.3200, value: 85000, name: '琶洲', district: '海珠', line: '8号线' },
      { lat: 23.1080, lng: 113.2600, value: 78000, name: '中山八', district: '荔湾', line: '5号线' },
      // 8号线
      { lat: 23.0950, lng: 113.2750, value: 82000, name: '沙园', district: '海珠', line: '8号线' },
      { lat: 23.0880, lng: 113.3000, value: 75000, name: '东晓南', district: '海珠', line: '2号线' },
      // 21号线
      { lat: 23.1100, lng: 113.4500, value: 45000, name: '科学城', district: '黄埔', line: '21号线' },
      // APM线
      { lat: 23.1200, lng: 113.3300, value: 68000, name: '大剧院', district: '天河', line: 'APM线' },
    ],
    regions: [
      { district: '天河', value: 185 },
      { district: '越秀', value: 162 },
      { district: '海珠', value: 128 },
      { district: '荔湾', value: 105 },
      { district: '白云', value: 88 },
      { district: '番禺', value: 72 },
      { district: '黄埔', value: 45 },
      { district: '南沙', value: 18 },
      { district: '增城', value: 22 },
      { district: '花都', value: 20 },
      { district: '从化', value: 15 },
    ],
  },
  chartData: {
    type: 'bar',
    title: '广州地铁日均客流量（万人次/区）',
    xField: 'district',
    yField: 'value',
    unit: '万人次',
  },
  stats: [
    { label: '运营线路', value: '16 条' },
    { label: '日均客流', value: '910 万人次' },
    { label: '最忙线路', value: '3号线' },
  ],
};

export default metro;

File: backend/app/crawlers/metro.py
"""Crawler: 广州地铁线路流量数据.

Data source: Publicly available metro station info + estimated ridership.
Runs standalone: python -m app.crawlers.metro
"""

from .base import BaseCrawler

# Public station-level data (coordinates from OpenStreetMap, ridership estimated from annual reports)
# In production, this would scrape from Guangzhou Metro official reports or open data portals.
STATIONS = [
    # Line 1
    {"name": "西塱", "lat": 23.065, "lng": 113.232, "line": "1号线/广佛线", "ridership": 85000, "district": "荔湾"},
    {"name": "坑口", "lat": 23.078, "lng": 113.231, "line": "1号线", "ridership": 35000, "district": "荔湾"},
    {"name": "花地湾", "lat": 23.087, "lng": 113.235, "line": "1号线", "ridership": 42000, "district": "荔湾"},
    {"name": "芳村", "lat": 23.098, "lng": 113.236, "line": "1号线", "ridership": 75000, "district": "荔湾"},
    {"name": "黄沙", "lat": 23.112, "lng": 113.240, "line": "1号线/6号线", "ridership": 68000, "district": "荔湾"},
    {"name": "长寿路", "lat": 23.118, "lng": 113.242, "line": "1号线", "ridership": 55000, "district": "荔湾"},
    {"name": "陈家祠", "lat": 23.120, "lng": 113.240, "line": "1号线/8号线", "ridership": 92000, "district": "荔湾"},
    {"name": "西门口", "lat": 23.125, "lng": 113.256, "line": "1号线", "ridership": 48000, "district": "越秀"},
    {"name": "公园前", "lat": 23.128, "lng": 113.264, "line": "1号线/2号线", "ridership": 128000, "district": "越秀"},
    {"name": "农讲所", "lat": 23.131, "lng": 113.275, "line": "1号线", "ridership": 62000, "district": "越秀"},
    {"name": "烈士陵园", "lat": 23.133, "lng": 113.281, "line": "1号线", "ridership": 55000, "district": "越秀"},
    {"name": "东山口", "lat": 23.132, "lng": 113.287, "line": "1号线/6号线", "ridership": 115000, "district": "越秀"},
    {"name": "杨箕", "lat": 23.128, "lng": 113.306, "line": "1号线/5号线", "ridership": 78000, "district": "越秀"},
    {"name": "体育西路", "lat": 23.131, "lng": 113.321, "line": "1号线/3号线", "ridership": 155000, "district": "天河"},
    {"name": "体育中心", "lat": 23.134, "lng": 113.326, "line": "1号线", "ridership": 82000, "district": "天河"},
    {"name": "广州东站", "lat": 23.150, "lng": 113.323, "line": "1号线/3号线", "ridership": 125000, "district": "天河"},
    # Line 3 (busiest)
    {"name": "番禺广场", "lat": 22.943, "lng": 113.359, "line": "3号线", "ridership": 72000, "district": "番禺"},
    {"name": "市桥", "lat": 22.950, "lng": 113.326, "line": "3号线", "ridership": 65000, "district": "番禺"},
    {"name": "汉溪长隆", "lat": 23.004, "lng": 113.332, "line": "3号线", "ridership": 58000, "district": "番禺"},
    {"name": "大石", "lat": 23.021, "lng": 113.317, "line": "3号线", "ridership": 70000, "district": "番禺"},
    {"name": "厦滘", "lat": 23.041, "lng": 113.318, "line": "3号线", "ridership": 45000, "district": "番禺"},
    {"name": "沥滘", "lat": 23.073, "lng": 113.318, "line": "3号线", "ridership": 62000, "district": "海珠"},
    {"name": "大塘", "lat": 23.089, "lng": 113.321, "line": "3号线", "ridership": 38000, "district": "海珠"},
    {"name": "客村", "lat": 23.100, "lng": 113.321, "line": "3号线/8号线", "ridership": 108000, "district": "海珠"},
    {"name": "广州塔", "lat": 23.108, "lng": 113.324, "line": "3号线/APM", "ridership": 78000, "district": "海珠"},
    {"name": "珠江新城", "lat": 23.120, "lng": 113.323, "line": "3号线/5号线", "ridership": 118000, "district": "天河"},
    {"name": "石牌桥", "lat": 23.134, "lng": 113.333, "line": "3号线", "ridership": 65000, "district": "天河"},
    {"name": "岗顶", "lat": 23.137, "lng": 113.340, "line": "3号线", "ridership": 75000, "district": "天河"},
    {"name": "华师", "lat": 23.143, "lng": 113.347, "line": "3号线", "ridership": 58000, "district": "天河"},
    {"name": "五山", "lat": 23.152, "lng": 113.352, "line": "3号线", "ridership": 42000, "district": "天河"},
    {"name": "天河客运站", "lat": 23.170, "lng": 113.345, "line": "3号线/6号线", "ridership": 88000, "district": "天河"},
    # Line 5
    {"name": "滘口", "lat": 23.115, "lng": 113.208, "line": "5号线", "ridership": 45000, "district": "荔湾"},
    {"name": "坦尾", "lat": 23.120, "lng": 113.221, "line": "5号线/6号线", "ridership": 32000, "district": "荔湾"},
    {"name": "中山八", "lat": 23.124, "lng": 113.234, "line": "5号线", "ridership": 78000, "district": "荔湾"},
    {"name": "西场", "lat": 23.132, "lng": 113.243, "line": "5号线", "ridership": 42000, "district": "荔湾"},
    {"name": "西村", "lat": 23.141, "lng": 113.248, "line": "5号线/8号线", "ridership": 55000, "district": "荔湾"},
    {"name": "广州火车站", "lat": 23.148, "lng": 113.256, "line": "2号线/5号线", "ridership": 110000, "district": "越秀"},
    {"name": "小北", "lat": 23.141, "lng": 113.277, "line": "5号线", "ridership": 52000, "district": "越秀"},
    {"name": "淘金", "lat": 23.138, "lng": 113.286, "line": "5号线", "ridership": 48000, "district": "越秀"},
    {"name": "区庄", "lat": 23.135, "lng": 113.298, "line": "5号线/6号线", "ridership": 65000, "district": "越秀"},
    {"name": "动物园", "lat": 23.138, "lng": 113.309, "line": "5号线", "ridership": 42000, "district": "越秀"},
    {"name": "猎德", "lat": 23.121, "lng": 113.333, "line": "5号线", "ridership": 112000, "district": "天河"},
    {"name": "潭村", "lat": 23.120, "lng": 113.346, "line": "5号线", "ridership": 38000, "district": "天河"},
    {"name": "员村", "lat": 23.118, "lng": 113.360, "line": "5号线/21号线", "ridership": 98000, "district": "天河"},
    {"name": "科韵路", "lat": 23.120, "lng": 113.371, "line": "5号线", "ridership": 45000, "district": "天河"},
    {"name": "车陂南", "lat": 23.118, "lng": 113.387, "line": "5号线/4号线", "ridership": 96000, "district": "天河"},
    {"name": "东圃", "lat": 23.115, "lng": 113.402, "line": "5号线", "ridership": 42000, "district": "天河"},
    {"name": "三溪", "lat": 23.112, "lng": 113.415, "line": "5号线", "ridership": 35000, "district": "天河"},
    {"name": "鱼珠", "lat": 23.105, "lng": 113.432, "line": "5号线/13号线", "ridership": 38000, "district": "黄埔"},
    {"name": "大沙地", "lat": 23.100, "lng": 113.446, "line": "5号线", "ridership": 32000, "district": "黄埔"},
    {"name": "大沙东", "lat": 23.097, "lng": 113.459, "line": "5号线", "ridership": 28000, "district": "黄埔"},
    {"name": "文冲", "lat": 23.094, "lng": 113.472, "line": "5号线", "ridership": 25000, "district": "黄埔"},
]

DISTRICT_RIDERSHIP = {
    "天河": 185, "越秀": 162, "海珠": 128, "荔湾": 105,
    "白云": 88, "番禺": 72, "黄埔": 45, "南沙": 18,
    "增城": 22, "花都": 20, "从化": 15,
}


class MetroCrawler(BaseCrawler):
    data_item_id = "metro"
    module = "pulse"

    def crawl(self) -> dict:
        """Return structured metro data. In production, this would scrape live data."""
        points = []
        for s in STATIONS:
            points.append({
                "lat": s["lat"],
                "lng": s["lng"],
                "value": s["ridership"],
                "name": s["name"],
                "district": s["district"],
                "line": s["line"],
            })

        regions = [
            {"district": d, "value": v}
            for d, v in DISTRICT_RIDERSHIP.items()
        ]

        return {
            "name": "地铁线路流量",
            "description": "广州地铁各线路主要站点日均客流量（基于公开年报估算）",
            "map_type": "heatmap",
            "chart_type": "bar",
            "chart_title": "广州地铁日均客流量（万人次/区）",
            "chart_unit": "万人次",
            "points": points,
            "regions": regions,
            "stats": [
                {"label": "运营线路", "value": "16 条"},
                {"label": "日均客流", "value": "910 万人次"},
                {"label": "最忙线路", "value": "3号线"},
            ],
        }


if __name__ == "__main__":
    crawler = MetroCrawler()
    try:
        crawler.run()
    finally:
        crawler.close()
